### PR TITLE
Refactor test_response to use Requests instead of curl

### PR DESCRIPTION
**Pull request by SlouchyButton**
I think we should prefer native-python approach when available. This will result in more clear code.

Also, we don't have to have two requests to check code and text, we can check both in one response.

### CODE REVIEW

File: container_ci_suite/engines/s2i_container.py [hunk 1]
```diff
@@ -464,49 +465,48 @@ def doc_content_old(self, strings: List) -> bool:
...
+        """
+
+        url = f"http://{url}:{port}"
         print(f"URL address to get response from container: {url}")
-        cmd_to_run = "curl --connect-timeout 10 -k -s -w '%{http_code}' " + f"{url}"
-        # Check if application returns proper HTTP_CODE
-        print("Check if HTTP_CODE is valid.")
-        for count in range(max_tests):
+
+        try:
```
Comment (phracek) [inline container_ci_suite/engines/s2i_container.py:488]: This is not needed.

Before 'while' set response to None and let's go to while cycle.
Also return value is boolean.

The code 
```
(response is not None and
               response.status_code != expected_code and
               (response.text != expected_output or expected_output == "") and
               max_tests > 0)
```
is duplicated. In case all conditions are met, then return False.
Let's make the condition not onliner.

File: container_ci_suite/engines/s2i_container.py [hunk 2]
```diff
@@ -464,49 +465,48 @@ def doc_content_old(self, strings: List) -> bool:
...
+            response = requests.get(url, timeout=10)
+        except requests.exceptions.ConnectionError:
+            response = None
+
```
Comment (phracek) [inline container_ci_suite/engines/s2i_container.py:492]: I would define here: correct_reposnse: bool = False

File: container_ci_suite/engines/s2i_container.py [hunk 3]
```diff
@@ -464,49 +465,48 @@ def doc_content_old(self, strings: List) -> bool:
...
-            if expected_output in output_code:
-                print(f"Expected output '{expected_output}' is present.")
-                return True
-            print(
-                f"check_response_inside_cluster:"
-                f"expected_output {expected_output} not found in output of {cmd_to_run} command. See {output_code}"
-            )
-            time.sleep(5)
-        return False
+                response = requests.get(url, timeout=10)
```
Comment (phracek) [inline container_ci_suite/engines/s2i_container.py:502]: It would be nice to inform users which conditions are met and which are not. Otherwise, it is hard to analyze the problem.

**Review by phracek — CHANGES_REQUESTED** [review]
Thanks for this pull request!!!

Please address my changes.

File: container_ci_suite/engines/s2i_container.py (hunk 1 above)
Comment (SlouchyButton) [reply container_ci_suite/engines/s2i_container.py]: Conditions were complicated, so I used simpler if statements to control the while loop.

Try/catch is needed because when request timeouts, it throws an exception. If we don't catch it, we fail the test immediately.

File: container_ci_suite/engines/s2i_container.py (hunk 3 above)
Comment (SlouchyButton) [reply container_ci_suite/engines/s2i_container.py:493]: I think that showing both code and text of the response is preferred, as the text can show important error message even if we have invalid code. Both are imo important together.

Mentioned this in comment.

**Comment by SlouchyButton** [comment]
One thing that also came to mind is whether we might want to throw error instead of return false. Throwing error will get caught by PyTest directly without us having to deal with it in the test.

**Comment by phracek** [comment]
> @phracek One thing that also came to mind is whether we might want to throw error instead of return false. Throwing error will get caught by PyTest directly without us having to deal with it in the test.

This could be problematic. E.g. In case we will call <API>.test_response and it failed with traceback, then we should do something like:
```
with pytest.raise as ConnectionError:
    <API>.test_response
```
which could be a problematic.

The return value True / False is for assert easy.

File: container_ci_suite/engines/s2i_container.py [hunk 4]
```diff
@@ -464,49 +465,62 @@ def doc_content_old(self, strings: List) -> bool:
...
-        url = f"{url}:{port}"
-        print(f"URL address to get response from container: {url}")
-        cmd_to_run = "curl --connect-timeout 10 -k -s -w '%{http_code}' " + f"{url}"
-        # Check if application returns proper HTTP_CODE
-        print("Check if HTTP_CODE is valid.")
-        for count in range(max_tests):
+        """
+        Test HTTP response of specified url
+
+        If expected output is empty, check will be skipped
```
Comment (phracek) [inline container_ci_suite/engines/s2i_container.py:475]: It does not make sense to call a response in case we do not want to check.
We always want to check an output to know the connection works properly.

File: container_ci_suite/engines/s2i_container.py [hunk 5]
```diff
@@ -464,49 +465,62 @@ def doc_content_old(self, strings: List) -> bool:
...
+                if response.status_code != expected_code:
+                    checks_passed = False
+                if (response.text != expected_output and
+                        expected_output != ""):
+                    checks_passed = False
+
+                # Log both code and text together as text can show error
+                # message
+                if not checks_passed:
+                    logger.error("Unexpected code %d or output %s, \
```
Comment (phracek) [inline container_ci_suite/engines/s2i_container.py:515]: Please use also "f" output like
logger.error(f"Unexpected code {response.status_code} or output {response.text}....)

**Review by phracek — CHANGES_REQUESTED** [review]
Last two issues. Otherwise LGTM. We are more or less done. Thanks for your engagement.

File: container_ci_suite/engines/s2i_container.py (hunk 4 above)
Comment (SlouchyButton) [reply container_ci_suite/engines/s2i_container.py:475]: Imagine that we use some external library (which we do in case of ruby, for example) as a test sample app. We might want to check that the app works - code 200 is returned, and content might be irrelevant for us as it can change without our knowledge.

I feel like having an option to pass the test only based on correct HTTP response code ignoring content is worth having (and even preferred for test stability).

> to know the connection works properly

If it is not, you won't get 200 HTTP code.

This feature of API is optional, so it is up to test developer to correctly assess whether checking response body is needed.

**Comment by SlouchyButton** [comment]
> > @phracek One thing that also came to mind is whether we might want to throw error instead of return false. Throwing error will get caught by PyTest directly without us having to deal with it in the test.
> 
> This could be problematic. E.g. In case we will call .test_response and it failed with traceback, then we should do something like:
> 
> ```
> with pytest.raise as ConnectionError:
>     <API>.test_response
> ```
> 
> which could be a problematic.
> 
> The return value True / False is for assert easy.

In the rare case where we _want_ the request to fail (which IDK if we even need anywhere as of now?) you are right that we would have to uses `pytest.raise`. On the other hand, I think that if we would assert in the `test_response` directly, it allows PyTest to work its magic of showing verbosity on demand. We wouldn't have to rely on logging as much, and we could debug test easier by just increasing verbosity. PyTest can show you exact differences and whole values of compared variables when you increase verbosity.

Both approaches seem valid, tho one uses PyTest internal functionality more than the other.

See further reading: https://docs.pytest.org/en/stable/how-to/output.html

File: container_ci_suite/engines/s2i_container.py (hunk 5 above)
Comment (SlouchyButton) [reply container_ci_suite/engines/s2i_container.py:515]: https://pylint.readthedocs.io/en/latest/user_guide/messages/warning/logging-fstring-interpolation.html
https://docs.python.org/3/howto/logging.html#optimization

TL;DR It is preferred to use '%s formatting' in logging functions as it allows logger to parse everything lazily - so it is preferred for optimization purposes.
